tapiriik/sync/sync.py

```diff
@@ -1148,6 +1148,10 @@ def Run(self, exhaustive=False, null_next_sync_on_unlock=False, heartbeat_callba
                     finally:
                         del activity
 
+                for conn in self._serviceConnections:
+                    srv = conn.Service
+                    srv.SynchronizationComplete(conn)

```

Comment:
This might not work in this location if Polar Flow expires their transactions after 10 minutes - first-time synchronization can easily take hours.

Reply:
Polar API does not support historical data so first-time sync will fetch no data.
Expired transaction _could_ happen if unpaid user would not trigger sync process for a very long time.
Anyway, if Polar expires transaction we will raise an exception during download routine and fall outside SynchronizationComplete code (nothing to commit). After that we will take another attempt to fetch _same_ polar data considering we already have uploaded some portion of it to other services.

Reply:
OK, I think this area might need some more work then - it would be best to create, fill, and commit transactions as many times as required. The site already has enough trouble with rate limits etc. necessitating multiple attempts to fully synchronize an account, so it'd be best to avoid introducing another such case by design.

Reply:
I agree. But I am not sure this is even possible using current architecture with separate list and download processes. I think we can handle it by automated sync relaunch skipping other services. Can be done by SynchronizationComplete function in case there is more data available. What do you think?

BTW. My previous implementation was downloading polar data in scope of activities listing action commiting and statring as many transactions as required. But this approach looks much more inappropriate.